src/wallet/hdchain.h

```diff
@@ -0,0 +1,70 @@
+#ifndef ZCOIN_HDCHAIN_H
+#define ZCOIN_HDCHAIN_H
+
+#include "key.h"
+
+/* simple HD chain data model */
+class CHDChain
+{
+public:
+    uint32_t nExternalChainCounter; // VERSION_BASIC
+    vector<uint32_t> nExternalChainCounters; // VERSION_WITH_BIP44: vector index corresponds to account value
+    SecureVector seed;
+    SecureVector mnemonic;
+    SecureVector passPhrase;
+    bool fIsCrypted;
+    bool f12Words;
+    CKeyID masterKeyID; //!< master key hash160
+
+    static const int VERSION_BASIC = 1;
+    static const int VERSION_WITH_BIP44 = 10;
+    static const int VERSION_WITH_BIP39 = 11;
+    static const int CURRENT_VERSION = VERSION_WITH_BIP39;
+    static const int N_CHANGES = 4; // standard = 0/1, mint = 2, exodus = 3
+    int nVersion;
+
+    CHDChain() { SetNull(); }
+    ADD_SERIALIZE_METHODS;
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion)
+    {
+
+        READWRITE(this->nVersion);
+        nVersion = this->nVersion;
+        READWRITE(nExternalChainCounter);
+        READWRITE(masterKeyID);
+        if(this->nVersion >= VERSION_WITH_BIP44){
+            READWRITE(nExternalChainCounters);
+            nExternalChainCounters.resize(N_CHANGES);
+        }
+        if(this->nVersion >= VERSION_WITH_BIP39) {
+            READWRITE(mnemonic);
+            READWRITE(passPhrase);
+            READWRITE(seed);
+            READWRITE(fIsCrypted);
+            READWRITE(f12Words);
+        }
```

Comment:
Why do we save/restore mnemonic, passPhrase, seed etc? Do we really really store password to the wallet in plain text next to it?

Reply:
we save it to be able to allow user to dump it any time.

Reply:
no! we might print it *once* during new wallet initialization. Sorry to overlook it earlier but there shouldn't be postfactum dumpmnemonic command at all. Look how it's done when setting up Trezor one for example.
The way you do it effectively ruins every wallet security. Only flags/counters/key ids can be saved

Reply:
I saw that's how dash is doing it but that's definitely not how it should be done

Reply:
Got to the bottom of it. Dash get its own implementation of `CHDChain` and it's encrypted/decrypted when saved into the `wallet.db`. Your implementation is a mix between that of bitcoin and dash. And that means all of these fields are saved to the wallet.db **unencrypted**. Which means a huge security hole

Reply:
It is being encrypted when you encrypt the wallet, it is being encrypted in our case also.

Reply:
nope, only keys are encrypted, everything you write to the wallet with `Write` is not